### PR TITLE
PR: Add code coverage report

### LINKED ISSUE #33: Extend plugin to compute and display coverage
This is still some way off.

### PR DESCRIPTION
**Pull request by stevetracvc**
Adds some minimal coverage reporting to the verbose output. Currently tested with pytest, and I think it'll work with nose. I don't know how to do this for unittest itself.

It's optional, a checkbox in the config dialog box.

Fixes #33

### CODE REVIEW

**Comment by stevetracvc** [comment]
Right now it uses the current working directory. Should it use that or the tests folder? I can see both being reasonable, but if someone selects a subfolder for the tests to perform, then it might not work properly. Maybe use the project folder itself?

**Comment by jitseniesen** [comment]
> Right now it uses the current working directory. Should it use that or the tests folder? I can see both being reasonable, but if someone selects a subfolder for the tests to perform, then it might not work properly. Maybe use the project folder itself?

The nose test runner uses a file in a temporary directory to collect the results, so you could also use that: https://github.com/spyder-ide/spyder-unittest/blob/b788ad9bfe37cfcdb6600c973d96f228df8cc208/spyder_unittest/backend/runnerbase.py#L130-L131

**Comment by stevetracvc** [comment]
Oh I just got what you meant. You mean for where to store the coverage files, right? I was referring to which folder is passed for source filtering:

```
coverage reporting with distributed testing support:
  --cov=[SOURCE]        Path or package name to measure during execution (multi-allowed). Use --cov= to not do any source filtering and record
                        everything.
```

After using it for a little while, I think the best option for me is to use current_project_path first, then fall back to the unittest wdir config value.

**Comment by stevetracvc** [comment]
@jitseniesen I have no idea what caused those tests to fail, I can't duplicate on my machine :confused:

**Comment by stevetracvc** [comment]
@jitseniesen We figured out what's making those Linux tests fail, the fix is in my other PR #174. I figured I'll wait till that gets merged, then I'll update this PR since there will be some conflicts between these two.

**Comment by jitseniesen** [comment]
Nice!

**Comment by stevetracvc** [comment]
@jitseniesen ok I merged in the most recent master commit, and I also added the coverage results to the GUI. Kinda hacky, so I don't know if you'll like it. :smile:

**Comment by stevetracvc** [comment]
I got a little bored. Last commit might be too much for a large project, but it splits the coverage report into individual files' reports too. You can even double click to take you to the file. The message shows which line numbers are missing in the coverage.

**Comment by ccordoba12** [comment]
@jitseniesen, thoughts about this?

**Comment by jitseniesen** [comment]
I should be able to have a proper look tomorrow. The use of fake tests to transmit the coverage results is a bit hacky but I think it is fine for an initial implementation.

**Comment by jitseniesen** [comment]
I tried running this without installing pytest-cov. When coverage testing is turned on, then pytest fails (unsurprisingly). This is not handled gracefully: basically, it seems that the tests just run forever. Can we improve the user experience in this case?

One part is outside the scope of this PR, so I opened another issue for this: Issue #176. In the context of this PR, it would be good if we can grey out the option for coverage testing if pytest-cov is not installed, or give a specific error message telling the user to install pytest-cov, or something like that.

Also, the coverage results for every file displayed twice: once inside the "Total coverage" pseudo-test, and once in a pseudo-test for coverage for that specific file. Why is that?

**Comment by stevetracvc** [comment]
Good catch, I hadn't tried that. I'm pretty sure there are checks for the other libraries, so I'll see how you did it and add something to gray out the checkbox if not installed.

The total report was the first one I made. Then later added the individual files reports. I think total is useful to show the full coverage result, but it could probably do away with the full text report.

**Comment by stevetracvc** [comment]
You can click on the file report to take you to the file. Like I said, I got a little bored so I started playing with it more 😀
If you think it's better to have one line for just the total report, expandable to show the full report details, let me know.

**Comment by stevetracvc** [comment]
I could probably even parse out the first line number from the missing report, and take the user there

**Comment by jitseniesen** [comment]
> You can click on the file report to take you to the file.

That's neat. 

> I could probably even parse out the first line number from the missing report, and take the user there

That is probably even better.

I think the ideal would be to have only the total report, expandable to show the full report, and that when you click on one of the lines in the full reports it would take you to that file. But that requires some deeper changes in the code that opens the editor when you click on a test result, so I am happy to leave that for later and keep the coverage reports as you have them now.

**Comment by stevetracvc** [comment]
I do like that idea, but that's beyond what I have time for right now :smile: 

A few things:
1. Currently, I need the first coverage report (total) so that I can add the total coverage percentage to the summary (title). However, the text when expanded is mostly duplicated when compared to the individual files. So, should I just ditch the extra_text for the total report, and it just be its own line?
2. I was thinking there could be a selection of colors, based on how much coverage % a given file has. They'd need to not conflict with anything currently in use. And it might just end up being more confusing than it's worth. Thoughts?
3. What about text translation? It doesn't look like it translates automatically, so I need to translate when creating the TestResult objects, right? But honestly, nothing seems to be translating for me, everything is still in english, even things wrapped with `_( )`

**Comment by stevetracvc** [comment]
Also I think part of the issue is that function show_versions uses the python that Spyder was launched from, rather than the currently chosen one.

**Comment by jitseniesen** [comment]
> Currently, I need the first coverage report (total) so that I can add the total coverage percentage to the summary (title). However, the text when expanded is mostly duplicated when compared to the individual files. So, should I just ditch the extra_text for the total report, and it just be its own line?

Up to you, you can ditch extra_text for the total report or leave it as it is now.

> I was thinking there could be a selection of colors, based on how much coverage % a given file has. They'd need to not conflict with anything currently in use. And it might just end up being more confusing than it's worth. Thoughts?

I don't immediately have an idea on what colours to use, so let's park this for now (unless you have a suggestion).

> * What about text translation? It doesn't look like it translates automatically, so I need to translate when creating the TestResult objects, right? But honestly, nothing seems to be translating for me, everything is still in english, even things wrapped with `_( )`

It does not translate automatically, there are some additional steps that needs to be done and one of them is somebody translating the strings manually. To be honest, I don't understand the details, I just follow the procedure in https://github.com/spyder-ide/spyder/blob/master/RELEASE.md .

> Also I think part of the issue is that function show_versions uses the python that Spyder was launched from, rather than the currently chosen one.

Good point, this needs to be fixed elsewhere. Let me open a new issue for it: Issue #177.

**Review by jitseniesen — CHANGES_REQUESTED** [review]
Just some minor suggestions for you to look over; if you disagree with anything, we can of course discuss it.

After you address these, I am happy to merge so if you want to add something else before this PR is merged do tell me!

File: spyder_unittest/backend/pytestrunner.py
Comment (stevetracvc) [reply spyder_unittest/backend/pytestrunner.py:57]: :+1:

File: spyder_unittest/backend/tests/test_runnerbase.py
Comment (stevetracvc) [reply spyder_unittest/backend/tests/test_runnerbase.py:30]: That was just to get more test coverage :laughing: While working on this plugin, I'd run the tests and see the coverage reports. It always said that it was missing coverage on RunnerBase's three abstract methods.

File: spyder_unittest/backend/noserunner.py
Comment (stevetracvc) [reply spyder_unittest/backend/noserunner.py:48]: I figured it'd be easier to just completely remove nose AFTER this PR. I'm not sure why I disabled the checkbox for nose but left this part in. I'll remove it

File: spyder_unittest/widgets/configdialog.py
Comment (stevetracvc) [reply spyder_unittest/widgets/configdialog.py:119]: I followed the pattern used for the ok_button code. 

~I first set it to disabled. Then if pytest is chosen, it gets enabled. If unittest or nose is chosen, it gets unchecked AND disabled. If I comment this out, then the checkbox is initially enabled even before a framework is chosen. I guess it's fairly moot though, since if you check it and then choose nose or unittest, then the checkbox will get unchecked and disabled.~

I think I get your question now. OK I moved it up to where I create the checkbox :smile:

File: spyder_unittest/backend/tests/test_runnerbase.py
Comment (stevetracvc) [reply spyder_unittest/backend/tests/test_runnerbase.py:30]: technically not related to this PR, want me to remove it?

File: spyder_unittest/backend/noserunner.py
Comment (jitseniesen) [reply spyder_unittest/backend/noserunner.py]: I agree, but I would rather wait until v0.6 to remove nose instead of removing functionality between v0.5.0 and v0.5.1 . I opened #178 to remind me.

File: spyder_unittest/backend/tests/test_runnerbase.py
Comment (jitseniesen) [reply spyder_unittest/backend/tests/test_runnerbase.py:30]: Don't remove it, it's fine, I just did not understand why it was in this PR.

Comment (stevetracvc) [reply spyder_unittest/backend/tests/test_runnerbase.py:30]: the coverage report kept annoying me :smile:

**Review by jitseniesen — APPROVED** [review]
Looks good now, thanks for all your work!

**Pull request merged**
